### PR TITLE
Add: align DeepSeek V4 Pro MoE with native MXFP8

### PR DESCRIPTION
**Pull request by yanghaoran29**
- Expand routed MXFP4 checkpoint weights into the MXFP8 Cube ABI in a
  new mx_utils.py host bridge, preserving the packed E8M0 scales, and
  quantize the shared-expert weights natively to MXFP8.
- Quantize the normalized MoE input and the SwiGLU activation with
  pl.quant_mx, and use pl.matmul_mx for the routed and shared expert
  projections.
- Dispatch the FP8 activation and its E8M0 scale through byte windows
  in moe.py, restoring the MX tiles with pl.reinterpret_view.
- Tile the shared and routed expert work so the MX scale buffers keep
  per-tile dependencies and W2 overlaps later gate work.
- Update the layer, MTP, forward, checkpoint, and documentation ABIs,
  and cover the MXFP4 bridge and the dispatch layout with contract
  tests.

### CODE REVIEW

**Comment by yanghaoran29** [comment]
Re AI review threads on this PR:

- `examples/mx/*` platform CLI / independent golden — **obsolete**: the whole `examples/mx/` tree was removed from this PR (superseded by in-model expert wiring).
- `mx_utils.NIBBLE_LUT` mapping and `fp4_packed_to_nibble_indices` packed-axis unpack — **already fixed** on current head (`4b62059`); host round-trip checks pass locally.

No further code change needed for those comments on this PR.

File: examples/mx/matmul_mx_dynamic_offset.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+            return int(str(val).split(",")[0].strip())
+    return int(os.environ.get("PYPTO_DEVICE_ID", "0"))
+
+
+def main() -> None:
+    if not all(hasattr(torch, name) for name in ("float8_e4m3fn", "float8_e8m0fnu")):
+        raise SystemExit("torch MX dtypes (float8_e4m3fn, float8_e8m0fnu) required")
+
+    platform = os.environ.get("PYPTO_PLATFORM", "a5")
+    device_id = _device_id_from_env()
```
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply examples/mx/matmul_mx_dynamic_offset.py]: Obsolete — `examples/mx/` (including this file) was removed from the PR; expert kernels supersede the standalone examples.

File: examples/mx/mxfp4_to_mxfp8_lut.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    golden = torch.empty(COLS, dtype=torch.int16)
+    lut = torch.tensor(NIBBLE_LUT, dtype=torch.int16)
+    for j in range(PACKED_COLS):
+        b = int(packed[j].item())
+        lo = b & 0x0F
+        hi = (b >> 4) & 0x0F
+        indices[2 * j] = lo
+        indices[2 * j + 1] = hi
+        golden[2 * j] = lut[lo]
+        golden[2 * j + 1] = lut[hi]
```
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply examples/mx/mxfp4_to_mxfp8_lut.py]: Obsolete — `examples/mx/mxfp4_to_mxfp8_lut.py` was removed from the PR.

File: models/deepseek_v4_pro/mx_utils.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    0x18,
+    0x1C,
+    0x80,
+    0x84,
+    0x88,
+    0x8C,
+    0x90,
+    0x94,
+    0x98,
+    0x9C,
```
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply models/deepseek_v4_pro/mx_utils.py]: Fixed on current head: `NIBBLE_LUT` maps FP4 nibbles to the correct E4M3FN codes (e.g. nibble 1 → `0x30` / 0.5). Verified with `nibble_indices_to_fp8` against the MXFP4 magnitude table.

File: models/deepseek_v4_pro/mx_utils.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    """Unpack FP4 bytes to INT16 nibble indices; last dim expands ×2 (lo then hi)."""
+    import torch
+
+    packed_u8 = packed.contiguous().view(torch.uint8)
+    *lead, half = packed_u8.shape
+    lo = (packed_u8 & 0x0F).to(torch.int16)
+    hi = ((packed_u8 >> 4) & 0x0F).to(torch.int16)
+    out = torch.empty(*lead, half * 2, dtype=torch.int16)
+    out[..., 0::2] = lo
+    out[..., 1::2] = hi
```
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply models/deepseek_v4_pro/mx_utils.py]: Fixed on current head: `fp4_packed_to_nibble_indices` unpacks axis 0 (`[inn/2, out]` → `[inn, out]`), matching `gen_mxfp4_weight_kn`. Round-trip check passes locally.
